src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuProcessPdpCancelPaidServiceImpl.java

```diff
@@ -4,77 +4,151 @@
+                                extractService.cancelPayment(dv, processDate);
                             }
+                        } else {
+                            extractService.resetFromExtraction(dv);
                         }
-                    } catch (WorkflowException we) {
-                        throw new RuntimeException("Could not retrieve document #"+documentNumber, we);
                     }
-                } else {
-                    LOG.warn("processPdpCancels() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
-                    continue;
+                } catch (WorkflowException we) {
+                    throw new RuntimeException("Could not retrieve document #" + documentNumber, we);
                 }
+            } else {
+                LOG.warn("processPdpCancel() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
+                return;
             }
+        }
+
+        paymentGroupService.processCancelledGroup(paymentDetail.getPaymentGroup(), processDate);
+    }
 
-            paymentGroupService.processCancelledGroup(paymentDetail.getPaymentGroup(), processDate);
+    /**
+     * Default implementation uses most of the "while" loop contents from the ProcessPdpCancelPaidServiceImpl.processPdpPaids method,
+     * with additional CU-related changes as needed. This implementation also runs within its own transaction.
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    @Override
+    public void processPdpPaid(PaymentDetail paymentDetail, Date processDate) {
+        String documentTypeCode = paymentDetail.getFinancialDocumentTypeCode();
+        String documentNumber = paymentDetail.getCustPaymentDocNbr();
+
+        if (purchasingAccountsPayableModuleService.isPurchasingBatchDocument(documentTypeCode)) {
+            purchasingAccountsPayableModuleService.handlePurchasingBatchPaids(documentNumber, documentTypeCode, processDate);
+        } else {
+            PaymentSourceToExtractService<PaymentSource> extractService = getPaymentSourceToExtractService(paymentDetail);
+            if (extractService != null) {
+                try {
+                    PaymentSource dv = (PaymentSource) documentService.getByDocumentHeaderId(documentNumber);
+                    extractService.markAsPaid(dv, processDate);
+                } catch (WorkflowException we) {
+                    throw new RuntimeException("Could not retrieve document #" + documentNumber, we);
+                }
+            } else {
+                LOG.warn("processPdpPaids() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
+                return;
+            }
         }
-   }
+
+        paymentGroupService.processPaidGroup(paymentDetail.getPaymentGroup(), processDate);
+    }
+
+    protected CuProcessPdpCancelPaidService getProxiedProcessPdpCancelPaidService() {
```

Comment:
Hi Chad,
I am not sure I understand why this method is needed?

Reply:
If the service calls its new processPdpCancel() or processPdpPaid() methods internally, then it won't observe the custom "Transactional" annotation setup on those methods. In order for the appropriate transactional behavior to occur, I have to call the methods on the Spring-generated proxy that wraps around this service instance.

Reply:
Thank you Chad.